test/test_replace.py

```diff
@@ -0,0 +1,239 @@
+
+    other_arg: str = "bob"
+
+
+@dataclass
+class NestedDataclasses:
+    nested_arg: str = "nested_arg"
+    arg1: int = 0
+    arg2: float = 0.0
+
+
+@dataclass
+class OuterDataclass(TestSetup):
+    some_arg: str = "some_arg"
+    nested: NestedDataclasses = field(default_factory=NestedDataclasses)
+
+
+@pytest.mark.parametrize(
+    ("config_cls", "args", "changes"),
+    [
+        (A, "--a 2.0", {"a": 2.0}),
+        pytest.param(
+            A, "--b 1.0", {"b": 1.0}, marks=pytest.mark.xfail(reason="Unrecognized arguments")
+        ),
+        (B, "--b test", {"b": "test"}),
+        (AB, "--a_or_b a", {"a_or_b": "a"}),
+        (AB, "--a_or_b b --b foo", {"__subgroups__@a_or_b": "b", "a_or_b": {"b": "foo"}}),
+        (AB, "--integer_in_string 2", {"integer_in_string": "2"}),
+        pytest.param(
+            AB,
+            "--integer_in_string 2",
+            {"integer_in_string": "2", "integer_only_by_post_init": 2},
+            marks=pytest.mark.xfail(reason="Any field with init=False will raise ValueError"),
+        ),
+        (CD, "", {}),
+        pytest.param(
+            lambda: "str_obj",
+            "",
+            {},
+            marks=pytest.mark.xfail(reason="Raise TypeError if obj is not dataclass instances"),
+        ),
+        (
+            OuterDataclass,
+            "--nested_arg nested_arg_2 --arg1 1 --arg2 2.0",
+            {"nested": {"nested_arg": "nested_arg_2", "arg1": 1, "arg2": 2.0}},
+        ),
+    ],
+)
+def test_replace_nested_dict(config_cls: type, args: str, changes: dict):
+    config = config_cls()
+    config_replaced = replace(config, changes)
+    logger.info(config_replaced)
+    logger.info(config.setup(args))
+    assert config.setup(args) == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@pytest.mark.parametrize(
+    ("config_cls", "args", "changes"),
+    [
+        (CD, "--c_or_d d --d 1", {"__subgroups__@c_or_d": "d", "c_or_d.d": 1}),
+        (CD, "--c_or_d c --c True", {"__subgroups__@c_or_d": "c", "c_or_d.c": True}),
+        (
+            OuterDataclass,
+            "--some_arg some_arg_1 --nested_arg nested_arg_1",
+            {"some_arg": "some_arg_1", "nested.nested_arg": "nested_arg_1"},
+        ),
+    ],
+)
+def test_replace_flatten_dict(config_cls: type, args: str, changes: dict):
+    config = config_cls()
+    config_replaced = replace(config, changes)
+    assert config.setup(args) == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@dataclass
+class InnerClass:
+    arg1: int = 0
+    arg2: str = "foo"
+
+
+@dataclass(frozen=True)
+class OuterClass:
+    outarg: int = 1
+    nested: InnerClass = field(default_factory=InnerClass)
+
+
+def test_replace_outerdataclass():
+    changes_1 = {"outarg": 2, "nested.arg1": 1, "nested.arg2": "bar"}
+    changes_2 = {"outarg": 2, "nested": {"arg1": 1, "arg2": "bar"}}
+    c = OuterClass()
+    c1 = replace(c, changes_1)
+    c2 = replace(c, changes_2)
+    assert c1 == c2
```

Comment:
Here you could also check that the result is correct:
```suggestion
    assert c1 == c2 == OuterClass(nested=InnerClass(arg1=1, arg2="bar"))
```